File: 001/01_join_test.R
renv::activate()
rm(list = ls())
library(tidyverse)
library(lubridate)

d_transaction <- read_csv('dummy_transaction_data.csv', 
                          col_types = cols(ym6 = col_character(), 
                                           company_number = col_character(), 
                                           q = col_double()))
resp_master <- read_csv('dummy_resp_master.csv', 
                        col_types = cols(.default = col_character()))

t0 <- Sys.time()
d <- d_transaction %>% left_join(resp_master, by = 'company_number')
d %>% dim %>% print()
print(Sys.time() - t0)


t0 <- Sys.time()
d2 <- d %>% 
  group_by(company_number) %>% 
  summarise(n=n(), x1 = sum(q), x2 = min(q), x3 = max(q), x4 = mean(q))
d2 %>% dim %>% print()
print(Sys.time() - t0)


t0 <- Sys.time()
tmp_d <- d %>% 
  mutate(tmp_ym6 = ym6 %>% paste0('01') %>% ymd) %>% 
  mutate(tmp_ym6_prev = tmp_ym6 %m+% months(1))
d3 <- tmp_d %>% 
  left_join(tmp_d, 
            by = c('tmp_ym6'='tmp_ym6_prev', 'company_number'='company_number'))
d3 %>% dim %>% print
print(Sys.time() - t0)

File: 001/01_join_test.py
import pandas as pd 
import numpy as np 
import datetime
from datetime import datetime


d_transaction = pd.read_csv('dummy_transaction_data.csv', dtype={'ym6':object, 'company_number':object, 'q':np.double})
resp_master = pd.read_csv('dummy_resp_master.csv', dtype={'company_number':object, 'person_name':object, 'division':object})


t0 = datetime.now()
d = d_transaction.merge(resp_master, on='company_number')
print(d.shape)
print(datetime.now() - t0)

t0 = datetime.now()
d2 = d.\
    groupby(['company_number']).\
        agg(n=('q', 'count'), x1=('q', 'sum'), x2=('q', 'min'), x3=('q', 'max'), x4=('q', 'mean'))
print(d2.shape)
print(datetime.now() - t0)

t0 = datetime.now()
d['tmp_ym6'] = pd.to_datetime(d['ym6']+'01')
d['tmp_ym6_prev'] = d['tmp_ym6'] + pd.tseries.offsets.DateOffset(months=1)
d3 = d.merge(d, left_on=['tmp_ym6_prev', 'company_number'], right_on=['tmp_ym6', 'company_number'])
print(d3.shape)
print(datetime.now() - t0)
